freqtrade/freqai/data_kitchen.py

```diff
@@ -0,0 +1,1308 @@
+import copy
+import datetime
+import json
+import logging
+import shutil
+from pathlib import Path
+from typing import Any, Dict, List, Tuple
+
+import numpy as np
+import numpy.typing as npt
+import pandas as pd
+from joblib import dump, load  # , Parallel, delayed # used for auto distribution assignment
+from joblib.externals import cloudpickle
+from pandas import DataFrame
+from sklearn import linear_model
+from sklearn.metrics.pairwise import pairwise_distances
+from sklearn.model_selection import train_test_split
+
+from freqtrade.configuration import TimeRange
+from freqtrade.data.history import load_pair_history
+from freqtrade.data.history.history_utils import refresh_backtest_ohlcv_data
+from freqtrade.exceptions import OperationalException
+from freqtrade.freqai.data_drawer import FreqaiDataDrawer
+from freqtrade.resolvers import ExchangeResolver
+from freqtrade.strategy.interface import IStrategy
+
+
+SECONDS_IN_DAY = 86400
+SECONDS_IN_HOUR = 3600
+
+logger = logging.getLogger(__name__)
+
+
+class FreqaiDataKitchen:
+    """
+    Class designed to analyze data for a single pair. Employed by the IFreqaiModel class.
+    Functionalities include holding, saving, loading, and analyzing the data.
+
+    Record of contribution:
+    FreqAI was developed by a group of individuals who all contributed specific skillsets to the
+    project.
+
+    Conception and software development:
+    Robert Caulk @robcaulk
+
+    Theoretical brainstorming:
+    Elin Törnquist @thorntwig
+
+    Code review, software architecture brainstorming:
+    @xmatthias
+
+    Beta testing and bug reporting:
+    @bloodhunter4rc, Salah Lamkadem @ikonx, @ken11o2, @longyu, @paranoidandy, @smidelis, @smarm
+    Juha Nykänen @suikula, Wagner Costa @wagnercosta
+    """
+
+    def __init__(
+        self,
+        config: Dict[str, Any],
+        data_drawer: FreqaiDataDrawer,
+        live: bool = False,
+        pair: str = "",
+    ):
+        self.data: Dict[Any, Any] = {}
+        self.data_dictionary: Dict[Any, Any] = {}
+        self.config = config
+        self.freqai_config = config["freqai"]
+        self.full_df: DataFrame = DataFrame()
+        self.append_df: DataFrame = DataFrame()
+        self.data_path = Path()
+        self.label_list: List = []
+        self.model_filename: str = ""
+        self.live = live
+        self.pair = pair
+        self.svm_model: linear_model.SGDOneClassSVM = None
+        self.keras = self.freqai_config.get("keras", False)
+        self.set_all_pairs()
+        if not self.live:
+            self.full_timerange = self.create_fulltimerange(
+                self.config["timerange"], self.freqai_config.get("train_period_days")
```

Comment:
i think we should add some "explicit" failure around backtesting without `--timerange`?
it's failing with keyerror if that's not given - and it's not obvious why it fails to the user...